p2p/channel.go

```diff
@@ -0,0 +1,55 @@
+package p2p
+
+import (
+	"github.com/gogo/protobuf/proto"
+)
+
+// ChannelID is an arbitrary channel ID.
+type ChannelID uint16
+
+// Envelope specifies the message receiver and sender.
+type Envelope struct {
+	From      PeerID        // Message sender, or empty for outbound messages.
+	To        PeerID        // Message receiver, or empty for inbound messages.
+	Broadcast bool          // Send message to all connected peers, ignoring To.
+	Message   proto.Message // Payload.
+}
+
+// Channel is a bidirectional channel for Protobuf message exchange with peers.
+type Channel struct {
+	// ID contains the channel ID.
+	ID ChannelID
+
+	// messageType specifies the type of messages exchanged via the channel, and
+	// is used e.g. for automatic unmarshaling.
+	messageType proto.Message
+
+	// In is a channel for receiving inbound messages. Envelope.From is always
+	// set.
+	In <-chan Envelope
+
+	// Out is a channel for sending outbound messages. Envelope.To or Broadcast
+	// must be set, otherwise the message is discarded.
+	Out chan<- Envelope
+
+	// Error is a channel for reporting peer errors to the router, typically used
+	// when peers send an invalid or malignant message.
+	Error chan<- PeerError
+}
+
+// Close closes the outbound channel, and is equivalent to close(Channel.Out).
+// This will cause Channel.In to be closed when appropriate. The ID can then be
+// reused.
+func (c *Channel) Close() error { return nil }
```

Comment:
This needs to be implemented. Do we close both channels?

Reply:
/cc @erikgrinaker

Reply:
Go channels must be closed by the sender - attempting to send on a closed channel panics, so the sender must coordinate it. This `Close()` method is intended for use by callers, thus it should close `Out` but not `In`.

`In` must be closed by whatever is sending on it when it is appropriate, so either `Router` or `Channel` depending on how we implement the internals. Typically, there's going to be some goroutine reading from the other side of `Out` - this should detect the closed `Out` channel after receiving all messages and then proceed to close `In` and stop sending on it.